subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -0,0 +1,2473 @@
+		len = snprintf(temp, sizeof(temp), "%s</%u>",
+			       (pos > 0) ? "," : "", obj->obj_id);
+		/*
+		 * TODO: iterate through resources once block transfer
+		 * is handled correctly
+		 */
+		if (pos + len < size) {
+			memcpy(&client_data[pos], temp, len);
+			pos += len;
+		} else {
+			/* full buffer -- exit loop */
+			break;
+		}
+
+		SYS_SLIST_FOR_EACH_CONTAINER(&engine_obj_inst_list,
+					     obj_inst, node) {
+			if (obj_inst->obj->obj_id == obj->obj_id) {
+				len = snprintf(temp, sizeof(temp),
+					       "%s</%s>",
+					       (pos > 0) ? "," : "",
+					       obj_inst->path);
+				/*
+				 * TODO: iterate through resources once block transfer
+				 * is handled correctly
+				 */
+				if (pos + len < size) {
+					memcpy(&client_data[pos], temp, len);
+					pos += len;
+				} else {
+					/* full buffer -- exit loop */
+					break;
+				}
+			}
+		}
+	}
+
+	client_data[pos] = '\0';
+	return pos;
+}
+
+/* input / output selection */
+
+static u16_t select_writer(struct lwm2m_output_context *out, u16_t accept)
+{
+	switch (accept) {
+
+	case LWM2M_FORMAT_PLAIN_TEXT:
+	case LWM2M_FORMAT_OMA_PLAIN_TEXT:
+		out->writer = &plain_text_writer;
+		break;
+
+	case LWM2M_FORMAT_OMA_TLV:
+	case LWM2M_FORMAT_OMA_OLD_TLV:
+		out->writer = &oma_tlv_writer;
+		break;
+
+#ifdef CONFIG_LWM2M_RW_JSON_SUPPORT
+	case LWM2M_FORMAT_OMA_JSON:
+	case LWM2M_FORMAT_OMA_OLD_JSON:
+		out->writer = &json_writer;
+		break;
+#endif
+
+	default:
+		SYS_LOG_ERR("Unknown Accept type %u, using LWM2M plain text",
+			    accept);
+		out->writer = &plain_text_writer;
+		accept = LWM2M_FORMAT_OMA_PLAIN_TEXT;
+		break;
+
+	}
+
+	return accept;
+}
+
+static u16_t select_reader(struct lwm2m_input_context *in, u16_t format)
+{
+	switch (format) {
+
+	case LWM2M_FORMAT_OMA_PLAIN_TEXT:
+		in->reader = &plain_text_reader;
+		break;
+
+	case LWM2M_FORMAT_OMA_TLV:
+	case LWM2M_FORMAT_OMA_OLD_TLV:
+		in->reader = &oma_tlv_reader;
+		break;
+
+	default:
+		SYS_LOG_ERR("Unknown content type %u, using LWM2M plain text",
+			    format);
+		in->reader = &plain_text_reader;
+		format = LWM2M_FORMAT_OMA_PLAIN_TEXT;
+		break;
+
+	}
+
+	return format;
+}
+
+/* check resource permission */
+int check_perm(u16_t obj_id, u16_t res_id, u8_t perm)
+{
+	struct lwm2m_engine_obj *obj;
+	struct lwm2m_engine_obj_field *obj_field;
+
+
+	obj = get_engine_obj(obj_id);
+	if (!obj) {
+		return 0;
+	}
+
+	obj_field = get_engine_obj_field(obj, res_id);
+	if (obj_field && ((obj_field->permissions & LWM2M_OP_BIT(perm)) ==
+			  LWM2M_OP_BIT(perm))) {
+		return 1;
+	}
+
+	return 0;
+}
+
+/* user data setter functions */
+
+static int string_to_path(char *pathstr, struct lwm2m_obj_path *path,
+			  char delim)
+{
+	u16_t value, len;
+	int i, begin = -1, toklen;
+
+	for (i = 0; i < strlen(pathstr); i++) {
+		/* skip non-numeric */
+		if (begin == -1 && pathstr[i] >= '0' && pathstr[i] <= '9') {
+			begin = i;
+		}
+
+		/* match delim or end of string */
+		if (begin >= 0 && (pathstr[i] == delim ||
```

Comment:
having a lot of for/if/else imbrication makes things hard to read. Though here it's contiki legacy I guess, could be worth to change it so if (<inverse of above condition>) { continue; }